File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/PurchaseResponseCallback.java
package in.dragonbra.javasteam.steam.handlers.steamapps.callback;

import in.dragonbra.javasteam.enums.EPurchaseResultDetail;
import in.dragonbra.javasteam.enums.EResult;
import in.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientPurchaseResponse;
import in.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg;
import in.dragonbra.javasteam.types.JobID;
import in.dragonbra.javasteam.types.KeyValue;
import in.dragonbra.javasteam.util.stream.MemoryStream;

import java.io.IOException;

public class PurchaseResponseCallback extends CallbackMsg {

    private EResult result;

    private EPurchaseResultDetail purchaseResultDetail;

    private KeyValue purchaseReceiptInfo;

    public PurchaseResponseCallback(JobID jobID, CMsgClientPurchaseResponse.Builder msg) {
        setJobID(jobID);

        this.result = EResult.from(msg.getEresult());
        this.purchaseResultDetail = EPurchaseResultDetail.from(msg.getPurchaseResultDetails());
        this.purchaseReceiptInfo = new KeyValue();

        if (msg.getPurchaseReceiptInfo() == null) {
            return;
        }

        try {
            MemoryStream ms = new MemoryStream(msg.getPurchaseReceiptInfo().toByteArray());
            this.purchaseReceiptInfo.tryReadAsBinary(ms);
        } catch (IOException exception) {
            throw new IllegalArgumentException("input stream is null");
        }
    }

    /**
     * @return Result of the operation
     */
    public EResult getResult() {
        return result;
    }

    /**
     * @return Purchase result of the operation
     */
    public EPurchaseResultDetail getPurchaseResultDetail() {
        return purchaseResultDetail;
    }

    /**
     * @return Purchase receipt of the operation
     */
    public KeyValue getPurchaseReceiptInfo() {
        return purchaseReceiptInfo;
    }
}

File: src/main/java/in/dragonbra/javasteam/steam/handlers/steamapps/callback/PurchaseResponseCallback.kt
package `in`.dragonbra.javasteam.steam.handlers.steamapps.callback

import `in`.dragonbra.javasteam.base.ClientMsgProtobuf
import `in`.dragonbra.javasteam.base.IPacketMsg
import `in`.dragonbra.javasteam.enums.EPurchaseResultDetail
import `in`.dragonbra.javasteam.enums.EResult
import `in`.dragonbra.javasteam.protobufs.steamclient.SteammessagesClientserver2.CMsgClientPurchaseResponse
import `in`.dragonbra.javasteam.steam.steamclient.callbackmgr.CallbackMsg
import `in`.dragonbra.javasteam.types.KeyValue
import `in`.dragonbra.javasteam.util.log.LogManager
import `in`.dragonbra.javasteam.util.stream.MemoryStream

/**
 * This callback is received in a response to activating a Steam key.
 */
@Suppress("MemberVisibilityCanBePrivate")
class PurchaseResponseCallback(packetMsg: IPacketMsg) : CallbackMsg() {

    companion object {
        private val logger = LogManager.getLogger(PurchaseResponseCallback::class.java)
    }

    /**
     * Gets Result of the operation
     */
    val result: EResult

    /**
     * Gets Purchase result of the operation
     */
    val purchaseResultDetail: EPurchaseResultDetail

    /**
     * Gets Purchase receipt of the operation
     */
    val purchaseReceiptInfo: KeyValue

    init {
        val purchaseResponse = ClientMsgProtobuf<CMsgClientPurchaseResponse.Builder>(
            CMsgClientPurchaseResponse::class.java,
            packetMsg
        )
        val msg = purchaseResponse.body

        jobID = purchaseResponse.targetJobID
        result = EResult.from(msg.eresult)
        purchaseResultDetail = EPurchaseResultDetail.from(msg.purchaseResultDetails)
        purchaseReceiptInfo = KeyValue()

        if (msg.purchaseReceiptInfo != null) {
            try {
                val ms = MemoryStream(msg.purchaseReceiptInfo.toByteArray())
                purchaseReceiptInfo.tryReadAsBinary(ms)
            } catch (e: Exception) {
                logger.error("Failed to read purchase receipt info", e)
            }
        }
    }
}
